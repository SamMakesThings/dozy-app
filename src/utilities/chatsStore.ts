import { create } from 'zustand';
import firestore from '@react-native-firebase/firestore';
import { Chat } from '../types/custom';

export type ChatsState = {
  chats: Chat[] | [];
  setChats: (chats: Chat[]) => void;
};

export const useChatsStore = create<ChatsState>((set) => ({
  chats: [
    {
      chatId: '',
      time: firestore.Timestamp.now(),
      message: '',
      sender: '',
      sentByUser: true,
    },
  ],
  setChats: (chats) =>
    set(() => {
      return { chats: chats };
    }),
}));
