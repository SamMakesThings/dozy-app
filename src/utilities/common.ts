export function convertAnalyticsEventName(originalName: string): string {
  const strWithUnderscore = originalName
    .replace(/[^a-z0-9\s_]/gi, '')
    .trim()
    .replace(/\s+_*|_+\s*/g, '_');

  return (
    /^[0-9]/.test(strWithUnderscore)
      ? `Event_${strWithUnderscore}`
      : strWithUnderscore
  ).substr(0, 40);
}
