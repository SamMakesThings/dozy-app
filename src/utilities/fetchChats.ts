import { FirebaseFirestoreTypes } from '@react-native-firebase/firestore';
import { Chat } from '../types/custom';

async function fetchChats(
  db: FirebaseFirestoreTypes.Module,
  userId: string,
): Promise<Chat[]> {
  // Retrieving chats from Firestore
  return new Promise((resolve) => {
    if (userId === undefined) {
      console.error('ERROR: userId passed to fetchChats is undefined');
    }

    // Pull the chats collection from Firestore, array it
    const colRef = db
      .collection('users')
      .doc(userId)
      .collection('supportMessages');

    colRef
      .orderBy('time', 'desc')
      .get()
      .then((res: FirebaseFirestoreTypes.QuerySnapshot) => {
        const chats: Array<Chat> = [];

        // Check that theres >1 entry. If no, set state accordingly
        if (res.size === 0) {
          resolve([]);
        }

        // Otherwise, arrange data and update state
        res.forEach(function (
          doc: FirebaseFirestoreTypes.QueryDocumentSnapshot,
        ) {
          const docData = doc.data();
          const currentChat = {
            sender: docData.sender,
            message: docData.message,
            sentByUser: docData.sentByUser,
            time: docData.time,
          };
          chats.push({ chatId: doc.id, ...currentChat });
        });
        resolve(chats);
      })
      .catch(function (error) {
        console.log('Error getting chats:', error);
      });
  });
}

export default fetchChats;
