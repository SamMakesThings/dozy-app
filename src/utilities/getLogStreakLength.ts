import { SleepLog } from '../types/custom';
import { subDays } from 'date-fns';

// Helper function. Return true if date2 is the day before date1 (or same day)
export function isPreviousDay(date1: Date, date2: Date): boolean {
  // Earliest candidate is midnight of 1 day ago
  const midnight = subDays(date1, 1);
  midnight.setHours(0, 0, 0, 0);
  return midnight <= date2 && date2 <= date1;
}

export function getLogStreakLength(sleepLogs: SleepLog[]): number {
  let streakLength = 0;

  // Calculate the current diary streak length
  while (true) {
    // Check that most recent log is yesterday or today. If not, break here at streak 0
    if (
      !sleepLogs.length ||
      !isPreviousDay(new Date(), sleepLogs[0].upTime.toDate())
    ) {
      break;
    }

    // Increment streak to start, since zero indexed
    streakLength++;

    // Check if current log day is the day after previous log day. If not, break
    if (
      sleepLogs.length === 1 || // If there's only 1 night logged, streak can be at most 1
      !sleepLogs[streakLength + 1] || // Make sure the next index is defined or the below code breaks
      !isPreviousDay(
        sleepLogs[streakLength].upTime.toDate(),
        sleepLogs[streakLength + 1].upTime.toDate(),
      )
    ) {
      break;
    }
  }

  return streakLength;
}
