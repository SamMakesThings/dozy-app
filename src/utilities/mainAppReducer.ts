export type AppState = {
  isLoading: boolean;
  isSigningIn: boolean;
  isSignout: boolean;
  userId: string | undefined;
  profileData: Record<string, any>;
  authLoading: boolean;
};

export const initialState: AppState = {
  isLoading: false,
  isSigningIn: false,
  isSignout: false,
  userId: undefined,
  profileData: {},
  authLoading: false,
};

export type ACTION =
  | {
      type: 'RESTORE_TOKEN';
      token: string;
      profileData: Record<string, any>;
    }
  | {
      type: 'SIGN_IN';
      token: string;
      onboardingComplete: boolean;
      profileData: Record<string, any>;
      userData: Record<string, any>;
    }
  | { type: 'SIGN_OUT' }
  | { type: 'SET_LOADING'; isLoading: boolean }
  | { type: 'SET_SIGNINGIN'; isSigningIn: boolean };

export const appReducer = (prevState: AppState, action: ACTION): AppState => {
  switch (action.type) {
    case 'RESTORE_TOKEN':
      return {
        ...prevState,
        userId: action.token,
        profileData: action.profileData,
      };
    case 'SIGN_IN':
      return {
        ...prevState,
        isSignout: false,
        userId: action.token,
        profileData: action.profileData,
      };
    case 'SIGN_OUT':
      return {
        ...prevState,
        isSignout: true,
        userId: undefined,
      };
    case 'SET_LOADING':
      return {
        ...prevState,
        isLoading: action.isLoading,
      };
    case 'SET_SIGNINGIN': {
      return {
        ...prevState,
        isSigningIn: action.isSigningIn,
      };
    }
  }
  return prevState;
};
