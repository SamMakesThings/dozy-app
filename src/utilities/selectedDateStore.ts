import { create } from 'zustand';

export type SelectedDateState = {
  selectedDate: {
    month: number;
    year: number;
    date: Date;
  };
  alterMonthSelection: (changeBy: number) => void;
};

export const useSelectedDateStore = create<SelectedDateState>((set) => ({
  month: new Date().getMonth(),
  selectedDate: {
    month: new Date().getMonth(),
    year: new Date().getFullYear(),
    date: new Date(),
  },
  alterMonthSelection: (changeBy: number) =>
    set((state) => {
      let { month, year, date } = state.selectedDate;
      const adjustedMonth = month + changeBy;

      if (adjustedMonth === 12) {
        month = 0;
        year += 1;
      } else if (adjustedMonth === -1) {
        month = 11;
        year += -1;
      } else {
        month = adjustedMonth;
      }

      date.setMonth(month, 15);
      date.setFullYear(year);

      return { selectedDate: { month: month, year: year, date: date } };
    }),
}));
