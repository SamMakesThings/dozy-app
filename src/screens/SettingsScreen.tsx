import React, { useCallback, useRef } from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  ScrollView,
} from 'react-native';
import {
  withTheme,
  ScreenContainer,
  Container,
  Icon,
  Switch,
  Touchable,
  DatePicker,
} from '@draftbit/ui';
import firestore from '@react-native-firebase/firestore';
import * as SecureStore from 'expo-secure-store';
import { scale } from 'react-native-size-matters';
import ExpoConstants from 'expo-constants';
import { take } from 'lodash';
import { dozy_theme } from '../config/Themes';
import Analytics from '../utilities/analytics.service';
import { encodeLocalTime, decodeServerTime } from '../utilities/time';
import Auth from '../utilities/auth.service';
import Notification from '../utilities/notification.service';
import AnalyticsEvents from '../constants/AnalyticsEvents';
import { useUserDataStore } from '../../src/utilities/userDataStore';

interface SettingsObject {
  logReminderTime: Date;
  logNotifsEnabled: boolean;
  targetBedTime: Date;
  targetWakeTime: Date;
  targetTimeInBed: number;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function SettingsScreen() {
  // Pass along the signOut function from the context provider
  const { state, signOut } = Auth.useAuth();
  const logReminderIdRef = useRef<string>();
  const theme = dozy_theme;

  const userData = useUserDataStore((userDataState) => userDataState.userData);
  const displayName = userData.userInfo.displayName;

  // Add function to update notification in Firebase based on local state
  const updateFbLogNotification = useCallback(
    (update) => {
      if (logReminderIdRef.current) {
        firestore()
          .collection('users')
          .doc(state.userId)
          .collection('notifications')
          .doc(logReminderIdRef.current)
          .update(update)
          .catch((err) => {
            console.log('error in notif write');
            console.log(err);
          });
      }
    },
    [state.userId],
  );

  // Update target sleep schedule values in currentTreatments in user object
  const updateTargetSleepSchedule = useCallback(
    async (update) => {
      if (state.userId) {
        try {
          await firestore().collection('users').doc(state.userId).set(
            {
              // testing: 'wassup',
              currentTreatments: update,
            },
            { merge: true },
          );
        } catch (err) {
          console.log(err);
        }
      } else {
        console.log("something isn't defined");
      }
    },
    [state.userId],
  );

  // Set up a reducer to manage settings state & keep Firebase updated
  // TODO: Have enabling notifs recheck permissions / Expo token
  const [settings, dispatch] = React.useReducer(
    (
      prevState: SettingsObject,
      action: { type: string; time?: Date; enabledStatus?: boolean },
    ): SettingsObject => {
      let encodedTimeData;

      switch (action.type) {
        case 'TOGGLE_LOG_NOTIFS':
          updateFbLogNotification({
            enabled: action.enabledStatus,
          });
          return {
            ...prevState,
            logNotifsEnabled: action.enabledStatus || true,
          };
        case 'SET_LOG_REMINDER_TIME':
          encodedTimeData = encodeLocalTime(action.time || new Date());
          updateFbLogNotification({
            time: encodedTimeData.value,
            version: encodedTimeData.version,
            timezone: encodedTimeData.timezone,
          });
          return {
            ...prevState,
            logReminderTime: action.time || new Date(),
          };
        case 'SET_TARGET_SLEEP_SCHEDULE':
          const targetWakeTime = action.time || new Date();
          const targetTimeInBed = userData?.currentTreatments?.targetTimeInBed;
          const targetBedTime =
            targetWakeTime && targetTimeInBed
              ? new Date(targetWakeTime.getTime() - targetTimeInBed * 60000)
              : new Date();

          updateTargetSleepSchedule({
            targetBedTime: targetBedTime,
            targetWakeTime: targetWakeTime,
            targetTimeInBed: targetTimeInBed,
          });

          return {
            ...prevState,
            targetBedTime: targetBedTime,
            targetWakeTime: targetWakeTime,
            targetTimeInBed: targetTimeInBed,
          };
        default:
          return prevState;
      }
    },
    {
      logReminderTime: new Date(),
      logNotifsEnabled: false,
      targetBedTime: new Date(),
      targetWakeTime: new Date(),
      targetTimeInBed: 480,
    },
  );

  const maybeAskNotificationPermission = useCallback(async () => {
    if (!(await Notification.isNotificationEnabled())) {
      const expoPushToken =
        await Notification.registerForPushNotificationsAsync(false, true);
      if (expoPushToken && state.userId) {
        Notification.updateExpoPushToken(expoPushToken, state.userId);
      }
    }
  }, [state.userId]);

  // Make sure the screen uses updated state once it loads for the first time.
  React.useEffect(() => {
    // Add function to pull existing settings from Firebase, update state with them
    const getSettings = async () => {
      const notifFbQuery = firestore()
        .collection('users')
        .doc(state.userId)
        .collection('notifications')
        .where('type', '==', 'DAILY_LOG');

      if (notifFbQuery === undefined) {
        return 'ERROR: Firebase not loading correctly';
      }

      const dailyLogNotificationDocs = await notifFbQuery.get();

      if (dailyLogNotificationDocs.docs.length) {
        const latestNotificationDoc =
          dailyLogNotificationDocs.docs[
            dailyLogNotificationDocs.docs.length - 1
          ];
        logReminderIdRef.current = latestNotificationDoc.id;
        latestNotificationDoc.ref.onSnapshot((notif) => {
          if (notif) {
            const notifData = notif.data();
            if (notifData) {
              dispatch({
                type: 'SET_LOG_REMINDER_TIME',
                time: decodeServerTime({
                  version: notifData.version,
                  value: notifData.time.toDate(),
                  timezone: notifData.timezone,
                }),
              });
              dispatch({
                type: 'TOGGLE_LOG_NOTIFS',
                enabledStatus: notifData.enabled,
              });
            }
          }
          // function setNewUserDataInState(newUserData: Record<string, any>) {

          // }
          // setNewUserDataInState(userData);
          // useUserDataStore.subscribe((newUserData) => {
          //   setNewUserDataInState(newUserData.userData);
          // });
        });
        // Delete notification docs except the latest one
        if (dailyLogNotificationDocs.docs.length > 1) {
          take(
            dailyLogNotificationDocs.docs,
            dailyLogNotificationDocs.docs.length - 1,
          ).forEach((it) => {
            it.ref.delete();
          });
          // Update logReminderId in user data
          firestore().collection('users').doc(state.userId).update({
            logReminderId: logReminderIdRef.current,
          });
        }
      }

      // Also fetch current target sleep schedule settings
      const currentTreatments = userData.currentTreatments;
      if (
        currentTreatments &&
        currentTreatments.targetBedTime &&
        currentTreatments.targetWakeTime &&
        currentTreatments.targetTimeInBed
      ) {
        const targetWakeTime =
          currentTreatments.targetWakeTime instanceof Date
            ? currentTreatments.targetWakeTime
            : new Date(currentTreatments.targetWakeTime.seconds * 1000);
        dispatch({
          type: 'SET_TARGET_SLEEP_SCHEDULE',
          time: targetWakeTime,
        });
      }
    };
    getSettings();
  }, []);

  return (
    <ScreenContainer
      hasSafeArea={true}
      scrollable={false}
      style={styles.Root_nd}
    >
      <ScrollView contentContainerStyle={styles.scrollViewContent}>
        <TouchableOpacity
          onPress={() => {
            if (!state.profileData.name) {
              const newProfileData = state.profileData;
              newProfileData.name = displayName || 'Temp Name';
              SecureStore.setItemAsync(
                'profileData',
                JSON.stringify(newProfileData),
              );
            }
          }}
          disabled={state.profileData.name || Platform.OS === 'android'}
        >
          <Container
            style={styles.Container_nz}
            elevation={0}
            useThemeGutterPadding={true}
          >
            <Icon
              name="Ionicons/ios-person"
              size={200}
              color={theme.colors.primary}
            />
            <Text
              style={[
                styles.Text_n1,
                theme.typography.headline3,
                {
                  color: theme.colors.strong,
                },
              ]}
            >
              {state.profileData.name || 'Tap here to fix chat'}
            </Text>
            <Text
              style={[
                styles.Text_nc,
                theme.typography.subtitle2,
                {
                  color: theme.colors.light,
                },
              ]}
              testID="appVersion"
            >
              {`@dozyapp ${ExpoConstants.nativeAppVersion}`}
            </Text>
          </Container>
        </TouchableOpacity>
        <Container
          style={styles.optionsWrapper}
          elevation={0}
          useThemeGutterPadding={true}
        >
          <Container
            style={[styles.optionsItem, styles.rowAlign]}
            elevation={0}
            useThemeGutterPadding={true}
          >
            <Text style={styles.optionsItemText}>
              Sleep log & check-in reminders
            </Text>
            <Switch
              color={theme.colors.primary}
              disabled={false}
              value={settings.logNotifsEnabled}
              onValueChange={async (value) => {
                dispatch({
                  type: 'TOGGLE_LOG_NOTIFS',
                  enabledStatus: value,
                });
                Analytics.logEvent(AnalyticsEvents.switchSleepLogReminders, {
                  enabled: value,
                });
                if (value) {
                  maybeAskNotificationPermission();
                }
              }}
            />
          </Container>
          <View style={styles.horizontalRule} />
          <Container
            style={styles.optionsItem}
            elevation={0}
            useThemeGutterPadding={true}
          >
            <Text style={styles.optionsItemText}>Sleep log reminder time</Text>
            <DatePicker
              style={styles.datePicker}
              mode="time"
              label="Time"
              type="underline"
              error={false}
              disabled={false}
              leftIconMode="inset"
              format="h:MM TT"
              date={settings.logReminderTime}
              onDateChange={(result) => {
                dispatch({ type: 'SET_LOG_REMINDER_TIME', time: result });
                Analytics.logEvent(AnalyticsEvents.editReminderTime);
                maybeAskNotificationPermission();
              }}
            />
          </Container>
          <View style={styles.horizontalRule} />
          <Container
            style={{
              ...styles.optionsItem,
              ...styles.optionsItemDisabledOpacity,
            }}
            elevation={0}
            useThemeGutterPadding={true}
          >
            <Text style={styles.optionsItemText}>Target bedtime</Text>
            <DatePicker
              style={styles.datePicker}
              mode="time"
              label="Time"
              type="underline"
              error={false}
              disabled={true}
              leftIconMode="inset"
              format="h:MM TT"
              date={settings.targetBedTime}
            />
          </Container>
          <View style={styles.horizontalRule} />
          <Container
            style={styles.optionsItem}
            elevation={0}
            useThemeGutterPadding={true}
          >
            <Text style={styles.optionsItemText}>Target wake time</Text>
            <DatePicker
              style={styles.datePicker}
              mode="time"
              label="Time"
              type="underline"
              error={false}
              disabled={false}
              leftIconMode="inset"
              format="h:MM TT"
              date={settings.targetWakeTime}
              onDateChange={(result) => {
                dispatch({ type: 'SET_TARGET_SLEEP_SCHEDULE', time: result });
                // updateTargetSleepSchedule({ targetWakeTime: new Date() });
                Analytics.logEvent(AnalyticsEvents.editTargetSleepSchedule);
              }}
            />
          </Container>
          <View style={styles.horizontalRule} />
          <Container
            style={{
              ...styles.optionsItem,
              ...styles.optionsItemDisabledOpacity,
            }}
            elevation={0}
            useThemeGutterPadding={true}
          >
            <Text style={styles.optionsItemText}>Target time in bed</Text>
            <Text style={styles.optionsItemText}>
              {settings.targetTimeInBed}
            </Text>
          </Container>
          <View style={styles.horizontalRule} />
          <Touchable style={styles.touchableItemWrapper} onPress={signOut}>
            <Container
              style={styles.optionsItem}
              elevation={0}
              useThemeGutterPadding={true}
            >
              <Text
                style={[styles.optionsItemText, { color: theme.colors.error }]}
              >
                Sign out of Dozy
              </Text>
            </Container>
          </Touchable>
        </Container>
      </ScrollView>
    </ScreenContainer>
  );
}

const styles = StyleSheet.create({
  scrollViewContent: {
    flexGrow: 1,
    justifyContent: 'space-around',
  },
  optionsItem: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    paddingVertical: 10,
    justifyContent: 'space-between',
  },
  optionsWrapper: {
    alignItems: 'flex-start',
    width: '100%',
    backgroundColor: dozy_theme.colors.medium,
    paddingVertical: 10,
    marginVertical: 50,
    borderRadius: 10,
    paddingLeft: 5,
    paddingRight: 0,
  },
  Container_nz: {
    alignItems: 'center',
    marginTop: scale(
      Platform.select({
        ios: dozy_theme.spacing.small,
        android: dozy_theme.spacing.medium,
      })!,
    ),
  },
  datePicker: {
    padding: 0,
    minWidth: 100,
    paddingVertical: 0,
    position: 'absolute',
    right: 0,
    top: -24,
  },
  Root_nd: {
    justifyContent: 'space-around',
    backgroundColor: dozy_theme.colors.background,
    paddingHorizontal: 20,
  },
  Text_n1: {
    textAlign: 'center',
    width: '100%',
  },
  Text_nc: {
    textAlign: 'center',
    width: '100%',
  },
  horizontalRule: {
    height: StyleSheet.hairlineWidth,
    width: '95%',
    backgroundColor: dozy_theme.colors.mediumInverse,
    alignSelf: 'flex-end',
    marginVertical: 10,
  },
  optionsItemText: {
    ...dozy_theme.typography.subtitle2,
    color: dozy_theme.colors.strong,
  },
  optionsItemDisabledOpacity: {
    opacity: 0.5,
  },
  touchableItemWrapper: {
    width: '100%',
  },
  rowAlign: {
    alignItems: 'center',
  },
});

export default withTheme(SettingsScreen);
