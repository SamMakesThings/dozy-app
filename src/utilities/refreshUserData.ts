import React from 'react';
import firestore from '@react-native-firebase/firestore';
import * as SecureStore from 'expo-secure-store';
import { ACTION } from './mainAppReducer';
import { useUserDataStore } from '../utilities/userDataStore';

export default async function refreshUserData(
  dispatch: React.Dispatch<ACTION>,
): Promise<void> {
  let userId;

  const updateUserData = useUserDataStore.getState().updateUserData;
  const setOnboardingComplete =
    useUserDataStore.getState().setOnboardingComplete;

  try {
    userId = await SecureStore.getItemAsync('userId');
    if (userId === null) {
      throw new Error("userId isn't stored in SecureStore, aborting");
    }
    // idToken = await SecureStore.getItemAsync('idToken');
    // providerId = await SecureStore.getItemAsync('providerId');
    const profileJson = await SecureStore.getItemAsync('profileData');

    // TODO: Add token validation
    if (profileJson) {
      const profileData = await JSON.parse(profileJson);
      dispatch({ type: 'RESTORE_TOKEN', token: userId, profileData });
    }

    // Update user's data from Firestore db
    firestore()
      .collection('users')
      .doc(userId)
      .get()
      .then((userData) => {
        updateUserData(userData.data() ?? {});
        userData.exists && userData.data()?.onboardingComplete === true
          ? setOnboardingComplete()
          : null;

        dispatch({ type: 'SET_LOADING', isLoading: false });
      });

    // Add a listener so user doc state updates live
    firestore()
      .collection('users')
      .doc(userId)
      .onSnapshot((userData) => {
        if (userData) {
          updateUserData(userData.data() ?? {});
          userData.exists && userData.data()?.onboardingComplete === true
            ? setOnboardingComplete()
            : null;
        } else {
          console.log("docSnapshot isn't defined at this point");
        }
      });
  } catch (e) {
    console.log('Error in restoring token:', e);
  }
}
