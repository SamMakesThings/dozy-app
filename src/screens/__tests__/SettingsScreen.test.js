import React from 'react';
import { shallow } from 'enzyme';
import * as Application from 'expo-application';
import { findByTestID } from '../../utilities/tests';
import { SettingsScreen } from '../SettingsScreen';

describe('SettingsScreen', () => {
  it('should match the snapshot', () => {
    // Mock new Date() to get consistent date prop for <DatePicker>
    const mockDate = new Date(Date.UTC(2021, 0, 1, 0));
    const dateSpy = jest
      .spyOn(global, 'Date')
      .mockImplementation(() => mockDate);

    const component = shallow(<SettingsScreen />);
    expect(component).toMatchSnapshot();

    dateSpy.mockRestore();
  });

  it('should show the app version correctly', () => {
    const component = shallow(<SettingsScreen />);
    expect(findByTestID(component, 'appVersion').props().children).toBe(
      `@dozyapp ${Application.nativeApplicationVersion}`,
    );
  });
});
