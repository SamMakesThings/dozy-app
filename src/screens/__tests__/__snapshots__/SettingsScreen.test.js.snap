// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen should match the snapshot 1`] = `
<withTheme(ScreenContainer)
  hasSafeArea={true}
  scrollable={false}
  style={
    Object {
      "backgroundColor": "rgba(35, 43, 63, 1)",
      "justifyContent": "space-around",
      "paddingHorizontal": 20,
    }
  }
>
  <ScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "justifyContent": "space-around",
      }
    }
  >
    <TouchableOpacity
      disabled={true}
      onPress={[Function]}
    >
      <withTheme(Container)
        elevation={0}
        style={
          Object {
            "alignItems": "center",
            "marginTop": NaN,
          }
        }
        useThemeGutterPadding={true}
      >
        <Icon
          color="rgba(0, 129, 138, 1)"
          name="Ionicons/ios-person"
          size={200}
        />
        <Text
          style={
            Array [
              Object {
                "textAlign": "center",
                "width": "100%",
              },
              Object {
                "fontFamily": "RubikMedium",
                "fontSize": 60,
                "lineHeight": 75,
              },
              Object {
                "color": "rgba(255, 255, 255, 1)",
              },
            ]
          }
        >
          Tap here to fix chat
        </Text>
        <Text
          style={
            Array [
              Object {
                "textAlign": "center",
                "width": "100%",
              },
              Object {
                "fontFamily": "RubikRegular",
                "fontSize": 30,
                "lineHeight": 47.14285714285714,
              },
              Object {
                "color": "rgba(153, 172, 185, 1)",
              },
            ]
          }
          testID="appVersion"
        >
          @dozyapp mock
        </Text>
      </withTheme(Container)>
    </TouchableOpacity>
    <withTheme(Container)
      elevation={0}
      style={
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "rgba(64, 75, 105, 1)",
          "borderRadius": 10,
          "marginVertical": 50,
          "paddingLeft": 5,
          "paddingRight": 0,
          "paddingVertical": 10,
          "width": "100%",
        }
      }
      useThemeGutterPadding={true}
    >
      <withTheme(Container)
        elevation={0}
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingVertical": 10,
            },
            Object {
              "alignItems": "center",
            },
          ]
        }
        useThemeGutterPadding={true}
      >
        <Text
          style={
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RubikRegular",
              "fontSize": 30,
              "lineHeight": 47.14285714285714,
            }
          }
        >
          Sleep log & check-in reminders
        </Text>
        <withTheme(Switch)
          color="rgba(0, 129, 138, 1)"
          disabled={false}
          onValueChange={[Function]}
          value={false}
        />
      </withTheme(Container)>
      <View
        style={
          Object {
            "alignSelf": "flex-end",
            "backgroundColor": "rgba(255, 255, 255, 0.87)",
            "height": 0.5,
            "marginVertical": 10,
            "width": "95%",
          }
        }
      />
      <withTheme(Container)
        elevation={0}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingVertical": 10,
          }
        }
        useThemeGutterPadding={true}
      >
        <Text
          style={
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RubikRegular",
              "fontSize": 30,
              "lineHeight": 47.14285714285714,
            }
          }
        >
          Sleep log reminder time
        </Text>
        <withTheme(DatePicker)
          date={2021-01-01T00:00:00.000Z}
          disabled={false}
          error={false}
          format="h:MM TT"
          label="Time"
          leftIconMode="inset"
          mode="time"
          onDateChange={[Function]}
          style={
            Object {
              "minWidth": 100,
              "padding": 0,
              "paddingVertical": 0,
              "position": "absolute",
              "right": 0,
              "top": -24,
            }
          }
          type="underline"
        />
      </withTheme(Container)>
      <View
        style={
          Object {
            "alignSelf": "flex-end",
            "backgroundColor": "rgba(255, 255, 255, 0.87)",
            "height": 0.5,
            "marginVertical": 10,
            "width": "95%",
          }
        }
      />
      <withTheme(Container)
        elevation={0}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 0.5,
            "paddingVertical": 10,
          }
        }
        useThemeGutterPadding={true}
      >
        <Text
          style={
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RubikRegular",
              "fontSize": 30,
              "lineHeight": 47.14285714285714,
            }
          }
        >
          Target bedtime
        </Text>
        <withTheme(DatePicker)
          date={2021-01-01T00:00:00.000Z}
          disabled={true}
          error={false}
          format="h:MM TT"
          label="Time"
          leftIconMode="inset"
          mode="time"
          style={
            Object {
              "minWidth": 100,
              "padding": 0,
              "paddingVertical": 0,
              "position": "absolute",
              "right": 0,
              "top": -24,
            }
          }
          type="underline"
        />
      </withTheme(Container)>
      <View
        style={
          Object {
            "alignSelf": "flex-end",
            "backgroundColor": "rgba(255, 255, 255, 0.87)",
            "height": 0.5,
            "marginVertical": 10,
            "width": "95%",
          }
        }
      />
      <withTheme(Container)
        elevation={0}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingVertical": 10,
          }
        }
        useThemeGutterPadding={true}
      >
        <Text
          style={
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RubikRegular",
              "fontSize": 30,
              "lineHeight": 47.14285714285714,
            }
          }
        >
          Target wake time
        </Text>
        <withTheme(DatePicker)
          date={2021-01-01T00:00:00.000Z}
          disabled={false}
          error={false}
          format="h:MM TT"
          label="Time"
          leftIconMode="inset"
          mode="time"
          onDateChange={[Function]}
          style={
            Object {
              "minWidth": 100,
              "padding": 0,
              "paddingVertical": 0,
              "position": "absolute",
              "right": 0,
              "top": -24,
            }
          }
          type="underline"
        />
      </withTheme(Container)>
      <View
        style={
          Object {
            "alignSelf": "flex-end",
            "backgroundColor": "rgba(255, 255, 255, 0.87)",
            "height": 0.5,
            "marginVertical": 10,
            "width": "95%",
          }
        }
      />
      <withTheme(Container)
        elevation={0}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 0.5,
            "paddingVertical": 10,
          }
        }
        useThemeGutterPadding={true}
      >
        <Text
          style={
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RubikRegular",
              "fontSize": 30,
              "lineHeight": 47.14285714285714,
            }
          }
        >
          Target time in bed
        </Text>
        <Text
          style={
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RubikRegular",
              "fontSize": 30,
              "lineHeight": 47.14285714285714,
            }
          }
        >
          480
        </Text>
      </withTheme(Container)>
      <View
        style={
          Object {
            "alignSelf": "flex-end",
            "backgroundColor": "rgba(255, 255, 255, 0.87)",
            "height": 0.5,
            "marginVertical": 10,
            "width": "95%",
          }
        }
      />
      <_default
        onPress={[Function]}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <withTheme(Container)
          elevation={0}
          style={
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingVertical": 10,
            }
          }
          useThemeGutterPadding={true}
        >
          <Text
            style={
              Array [
                Object {
                  "color": "rgba(255, 255, 255, 1)",
                  "fontFamily": "RubikRegular",
                  "fontSize": 30,
                  "lineHeight": 47.14285714285714,
                },
                Object {
                  "color": "rgba(255, 69, 100, 1)",
                },
              ]
            }
          >
            Sign out of Dozy
          </Text>
        </withTheme(Container)>
      </_default>
    </withTheme(Container)>
  </ScrollView>
</withTheme(ScreenContainer)>
`;
